src/core/AudioEngineProfiler.cpp

```diff
@@ -38,10 +41,23 @@ AudioEngineProfiler::AudioEngineProfiler() :
 
 void AudioEngineProfiler::finishPeriod( sample_rate_t sampleRate, fpp_t framesPerPeriod )
 {
-	int periodElapsed = m_periodTimer.elapsed();
+	// Time taken to process all data and fill the audio buffer.
+	const unsigned int periodElapsed = m_periodTimer.elapsed();
+	// Maximum time the processing can take before causing buffer underflow. Convert to us.
+	const uint64_t timeLimit = static_cast<uint64_t>(1000000) * framesPerPeriod / sampleRate;
+
+	// Compute new overall CPU load and apply exponential averaging.
+	// The result is used for overload detection in AudioEngine::CriticalXRuns()
+	// → the weight of a new sample must be high enough to allow relatively fast changes!
+	const int newCpuLoad = 100 * periodElapsed / timeLimit;
```

Comment:
This, and the detailed loads, may as well be computed using floating point arithmetic for extra precision, since the result is stored as a float too.

Reply:
It does not seem to make any difference; the bar graph still only moves in steps of 4 pixels:
https://github.com/LMMS/lmms/assets/20932751/7ddb3549-3413-49db-8024-360b966124a3

So the (admittedly tiny) performance hit would probably not serve any purpose, unless the CPU graph widget is updated to show to better resolution as well. And even after that, the `newCpuLoad` is multiplied by 0.1 on the next line, further reducing any impact the loss of precision brings.

But crunching a few floats per buffer instead of integers probably has no measurable performance impact, so I don't mind either way.

Reply:
Oh, a comment in `CPULoadWidget` reminded me: the low drawing resolution is probably intentional backward compatibility with the classic theme. But as far as I can tell, it's not working particularly well anyway:
![screen](https://github.com/LMMS/lmms/assets/20932751/b320de1b-5318-4c8f-9fbc-b46b7b9e08d7)

Maybe it's just because of my high DPI screen, but the drawing steps do not seem to properly align with the "LEDs", which was the sole purpose of the "rounding code". So if nobody objects, I would just get rid of it and let the load graph move smoothly.

Reply:
Even if the CPU graph precision is limited to some number of pixels, it could still be worth tracking the load more accurately behind the scene. And since we have a tooltip now, it's displayed to the nearest percentage point anyway.

As for a smoother resolution, I very much agree. I though we already had a pull request for that (possibly by @LostRobotMusic), but I can't seem to find one. If you really want to push the boat out, you could add a CSS property for the bar step size, so themes can set it appropriately for their own image.

Reply:
> could still be worth tracking the load more accurately behind the scene.

That's true; if someone decides to fix / improve things related to `AudioEngine::criticalXRuns()`, it's probably better to have numbers without unnecessary rounding as the input.

So, pushing a change to floats then. (And the smoother stepping; I also remember some old change made by Lost, but maybe it addressed something else rather than the drawing step.)

Reply:
> If you really want to push the boat out, you could add a CSS property for the bar step size

I took it as an opportunity to finally learn how styling and custom attributes in Qt work, and implemented the custom step size support.

While at it, I found out that the misalignment in classic theme was not caused by scaling. It persisted even with `QT_AUTO_SCREEN_SCALE_FACTOR=0`, because the old fixed code assumed LED = 2 pixels + 1 pixel space, while the classic theme uses 3+1.

So now the default theme moves smoother and the classic theme looks nicer. :)